File: backend/app/api/v2/web_functions.py
"""
Web 功能和子功能管理 API

提供 Web 功能和子功能的 CRUD 操作接口
"""
"""
版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。

本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。

授权商业应用请联系微信：huice666
"""


from typing import Optional

from fastapi import APIRouter, Query
from pydantic import BaseModel

from app.api.deps import (
    WebFunctionServiceDep,
    PaginationDep,
)
from app.schemas.common import SuccessResponse

# noqa  MC80OmFIVnBZMlhsdktEbHVwNDZNalJpV0E9PTo3YzAyZDMzMA==

# ============ 请求模型 ============

class CreateWebFunctionRequest(BaseModel):
    """创建 Web 功能请求"""
    display_name: str
    name: str
    folder_id: Optional[str] = None
    description: Optional[str] = None
    base_url: Optional[str] = None
    business_module: Optional[str] = None
    navigation: Optional[dict] = None
    pages: Optional[list] = None
    tags: Optional[list] = None
    custom_config: Optional[dict] = None
    sort_order: int = 0


class UpdateWebFunctionRequest(BaseModel):
    """更新 Web 功能请求"""
    display_name: Optional[str] = None
    name: Optional[str] = None
    description: Optional[str] = None
    base_url: Optional[str] = None
    business_module: Optional[str] = None
    navigation: Optional[dict] = None
    pages: Optional[list] = None
    tags: Optional[list] = None
    custom_config: Optional[dict] = None
    sort_order: Optional[int] = None


class CreateWebSubFunctionRequest(BaseModel):
    """创建 Web 子功能请求"""
    function_id: str
    display_name: str
    name: str
    folder_id: Optional[str] = None
    description: Optional[str] = None
    test_type: str = "functional"
    target_pages: Optional[list] = None
    test_scenario: Optional[str] = None
    test_data: Optional[dict] = None
    expected_results: Optional[list] = None
    priority: str = "medium"
    tags: Optional[list] = None
    custom_config: Optional[dict] = None
    sort_order: int = 0


class UpdateWebSubFunctionRequest(BaseModel):
    """更新 Web 子功能请求"""
    display_name: Optional[str] = None
    name: Optional[str] = None
    description: Optional[str] = None
    test_type: Optional[str] = None
    target_pages: Optional[list] = None
    test_scenario: Optional[str] = None
    test_data: Optional[dict] = None
    expected_results: Optional[list] = None
    priority: Optional[str] = None
    tags: Optional[list] = None
    custom_config: Optional[dict] = None
    sort_order: Optional[int] = None


# ============ 路由定义 ============

router = APIRouter(prefix="/projects/{project_identifier}/web-functions")


# ============ Web 功能管理接口 ============

@router.post(
    "",
    response_model=SuccessResponse,
    summary="创建 Web 功能",
    description="创建新的 Web 功能定义",
)
async def create_web_function(
    project_identifier: str,
    request: CreateWebFunctionRequest,
    service: WebFunctionServiceDep,
):
    """创建 Web 功能"""
    result = await service.create_web_function(
        project_identifier=project_identifier,
        **request.model_dump(),
    )
# noqa  MS80OmFIVnBZMlhsdktEbHVwNDZNalJpV0E9PTo3YzAyZDMzMA==

    return SuccessResponse(data=result, message="Web 功能创建成功")


@router.get(
    "",
    response_model=SuccessResponse,
    summary="获取 Web 功能列表",
    description="获取项目下的所有 Web 功能列表，支持搜索和过滤",
)
async def list_web_functions(
    project_identifier: str,
    service: WebFunctionServiceDep,
    pagination: PaginationDep,
    folder_id: Optional[str] = Query(None, description="文件夹 ID 过滤"),
    search: Optional[str] = Query(None, description="搜索关键词"),
):
    """获取 Web 功能列表"""
    result = await service.list_web_functions(
        project_identifier=project_identifier,
        folder_id=folder_id,
        offset=(pagination.p - 1) * pagination.page_size,
        limit=pagination.page_size,
        search=search,
    )

    return SuccessResponse(data=result)


# ============ Web 子功能管理接口 ============

@router.get(
    "/sub-functions",
    response_model=SuccessResponse,
    summary="获取 Web 子功能列表",
    description="获取项目下的所有 Web 子功能列表，支持搜索和过滤",
)
async def list_web_sub_functions(
    project_identifier: str,
    service: WebFunctionServiceDep,
    pagination: PaginationDep,
    function_id: Optional[str] = Query(None, description="功能 ID 过滤"),
    folder_id: Optional[str] = Query(None, description="文件夹 ID 过滤"),
    search: Optional[str] = Query(None, description="搜索关键词"),
):
    """获取 Web 子功能列表"""
    result = await service.list_web_sub_functions(
        project_identifier=project_identifier,
        function_id=function_id,
        folder_id=folder_id,
        offset=(pagination.p - 1) * pagination.page_size,
        limit=pagination.page_size,
        search=search,
    )

    return SuccessResponse(data=result)


@router.post(
    "/sub-functions",
    response_model=SuccessResponse,
    summary="创建 Web 子功能",
    description="创建新的 Web 子功能定义",
)
async def create_web_sub_function(
    project_identifier: str,
    request: CreateWebSubFunctionRequest,
    service: WebFunctionServiceDep,
):
    """创建 Web 子功能"""
    result = await service.create_web_sub_function(
        project_identifier=project_identifier,
        **request.model_dump(),
    )

    return SuccessResponse(data=result, message="Web 子功能创建成功")
# noqa  Mi80OmFIVnBZMlhsdktEbHVwNDZNalJpV0E9PTo3YzAyZDMzMA==


@router.get(
    "/sub-functions/{sub_function_id}",
    response_model=SuccessResponse,
    summary="获取 Web 子功能详情",
    description="获取指定 Web 子功能的详细信息",
)
async def get_web_sub_function(
    sub_function_id: str,
    service: WebFunctionServiceDep,
):
    """获取 Web 子功能详情"""
    result = await service.get_web_sub_function(sub_function_id=sub_function_id)
    return SuccessResponse(data=result)


@router.patch(
    "/sub-functions/{sub_function_id}",
    response_model=SuccessResponse,
    summary="更新 Web 子功能",
    description="更新指定 Web 子功能的信息",
)
async def update_web_sub_function(
    sub_function_id: str,
    request: UpdateWebSubFunctionRequest,
    service: WebFunctionServiceDep,
):
    """更新 Web 子功能"""
    result = await service.update_web_sub_function(
        sub_function_id=sub_function_id,
        **request.model_dump(exclude_none=True),
    )

    return SuccessResponse(data=result, message="Web 子功能更新成功")


@router.delete(
    "/sub-functions/{sub_function_id}",
    response_model=SuccessResponse,
    summary="删除 Web 子功能",
    description="删除指定的 Web 子功能",
)
async def delete_web_sub_function(
    sub_function_id: str,
    service: WebFunctionServiceDep,
):
    """删除 Web 子功能"""
    result = await service.delete_web_sub_function(sub_function_id=sub_function_id)
    return SuccessResponse(data=result)


@router.get(
    "/sub-functions/{sub_function_id}/artifacts",
    response_model=SuccessResponse,
    summary="获取 Web 子功能测试成果物",
    description="获取指定 Web 子功能的所有测试成果物（测试计划、测试用例、测试脚本、测试报告）",
)
async def get_web_sub_function_artifacts(
    sub_function_id: str,
    service: WebFunctionServiceDep,
):
    """获取 Web 子功能测试成果物"""
    result = await service.get_sub_function_artifacts(sub_function_id=sub_function_id)
    return SuccessResponse(data=result)


@router.get(
    "/{function_id}",
    response_model=SuccessResponse,
    summary="获取 Web 功能详情",
    description="获取指定 Web 功能的详细信息",
)
async def get_web_function(
    function_id: str,
    service: WebFunctionServiceDep,
):
    """获取 Web 功能详情"""
    result = await service.get_web_function(function_id=function_id)
    return SuccessResponse(data=result)
# pylint: disable  My80OmFIVnBZMlhsdktEbHVwNDZNalJpV0E9PTo3YzAyZDMzMA==


@router.patch(
    "/{function_id}",
    response_model=SuccessResponse,
    summary="更新 Web 功能",
    description="更新指定 Web 功能的信息",
)
async def update_web_function(
    function_id: str,
    request: UpdateWebFunctionRequest,
    service: WebFunctionServiceDep,
):
    """更新 Web 功能"""
    result = await service.update_web_function(
        function_id=function_id,
        **request.model_dump(exclude_none=True),
    )

    return SuccessResponse(data=result, message="Web 功能更新成功")


@router.delete(
    "/{function_id}",
    response_model=SuccessResponse,
    summary="删除 Web 功能",
    description="删除指定的 Web 功能",
)
async def delete_web_function(
    function_id: str,
    service: WebFunctionServiceDep,
):
    """删除 Web 功能"""
    result = await service.delete_web_function(function_id=function_id)
    return SuccessResponse(data=result)

File: ui/lib/api/web-functions.ts
/**
 * 版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。
 * 
 * 本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
 * 不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。
 * 
 * 授权商业应用请联系微信：huice666
 */
// eslint-disable  MC80OmFIVnBZMlhsdktEbHVwNDZSbkZxYVE9PTo0N2E2MDQ1ZA==

/**
 * Web 功能和子功能相关的 API 客户端函数
 */

import { apiClient } from "./client";

// ==================== 类型定义 ====================

export interface WebFunction {
  id: string;
  identifier: string;
  display_name: string;
  name: string;
  description: string | null;
  base_url: string | null;
  business_module: string | null;
  navigation: any;
  pages: any;
  tags: any;
  custom_config: any;
  folder_id: string | null;
  total_sub_functions: number;
  total_test_cases: number;
  total_test_runs: number;
  last_run_status: string | null;
  sort_order: number;
  created_at: string;
  updated_at: string | null;
}

export interface WebSubFunction {
  id: string;
  identifier: string;
  function_id: string;
  display_name: string;
  name: string;
  description: string | null;
  test_type: string;
  target_pages: any;
  test_scenario: string | null;
  test_data: any;
  expected_results: any;
  priority: string;
  tags: any;
  custom_config: any;
  folder_id: string | null;
  total_test_cases: number;
  total_test_runs: number;
  last_run_status: string | null;
  sort_order: number;
  created_at: string;
  updated_at: string | null;
}

export interface CreateWebFunctionRequest {
  display_name: string;
  name: string;
  folder_id?: string;
  description?: string;
  base_url?: string;
  business_module?: string;
  navigation?: any;
  pages?: any;
  tags?: any;
  custom_config?: any;
  sort_order?: number;
}
// FIXME  MS80OmFIVnBZMlhsdktEbHVwNDZSbkZxYVE9PTo0N2E2MDQ1ZA==

export interface UpdateWebFunctionRequest {
  display_name?: string;
  name?: string;
  description?: string;
  base_url?: string;
  business_module?: string;
  navigation?: any;
  pages?: any;
  tags?: any;
  custom_config?: any;
  sort_order?: number;
}

export interface CreateWebSubFunctionRequest {
  function_id: string;
  display_name: string;
  name: string;
  folder_id?: string;
  description?: string;
  test_type?: string;
  target_pages?: any;
  test_scenario?: string;
  test_data?: any;
  expected_results?: any;
  priority?: string;
  tags?: any;
  custom_config?: any;
  sort_order?: number;
}

export interface UpdateWebSubFunctionRequest {
  display_name?: string;
  name?: string;
  description?: string;
  test_type?: string;
  target_pages?: any;
  test_scenario?: string;
  test_data?: any;
  expected_results?: any;
  priority?: string;
  tags?: any;
  custom_config?: any;
  sort_order?: number;
}

export interface WebFunctionListResponse {
  items: WebFunction[];
  total: number;
  offset: number;
  limit: number;
}

export interface WebSubFunctionListResponse {
  items: WebSubFunction[];
  total: number;
  offset: number;
  limit: number;
}

// ==================== Web 功能 API 函数 ====================

/**
 * 获取项目的 Web 功能列表
 */
export async function listWebFunctions(
  projectIdentifier: string,
  params?: {
    p?: number;
    page_size?: number;
    folder_id?: string;
    search?: string;
  }
): Promise<WebFunctionListResponse> {
  const response = await apiClient.get<{ data: WebFunctionListResponse }>(
    `/projects/${projectIdentifier}/web-functions`,
    { params }
  );
  return response.data;
}
// @ts-expect-error  Mi80OmFIVnBZMlhsdktEbHVwNDZSbkZxYVE9PTo0N2E2MDQ1ZA==

/**
 * 获取 Web 功能详情
 */
export async function getWebFunction(
  projectIdentifier: string,
  functionId: string
): Promise<WebFunction> {
  const response = await apiClient.get<{ data: WebFunction }>(
    `/projects/${projectIdentifier}/web-functions/${functionId}`
  );
  return response.data;
}

/**
 * 创建 Web 功能
 */
export async function createWebFunction(
  projectIdentifier: string,
  data: CreateWebFunctionRequest
): Promise<WebFunction> {
  const response = await apiClient.post<{ data: WebFunction }>(
    `/projects/${projectIdentifier}/web-functions`,
    data
  );
  return response.data;
}

/**
 * 更新 Web 功能
 */
export async function updateWebFunction(
  projectIdentifier: string,
  functionId: string,
  data: UpdateWebFunctionRequest
): Promise<WebFunction> {
  const response = await apiClient.patch<{ data: WebFunction }>(
    `/projects/${projectIdentifier}/web-functions/${functionId}`,
    data
  );
  return response.data;
}

/**
 * 删除 Web 功能
 */
export async function deleteWebFunction(
  projectIdentifier: string,
  functionId: string
): Promise<{ success: boolean; message: string }> {
  const response = await apiClient.delete<{ data: { success: boolean; message: string } }>(
    `/projects/${projectIdentifier}/web-functions/${functionId}`
  );
  return response.data;
}

// ==================== Web 子功能 API 函数 ====================

/**
 * 获取项目的 Web 子功能列表
 */
export async function listWebSubFunctions(
  projectIdentifier: string,
  params?: {
    p?: number;
    page_size?: number;
    function_id?: string;
    folder_id?: string;
    search?: string;
  }
): Promise<WebSubFunctionListResponse> {
  const response = await apiClient.get<{ data: WebSubFunctionListResponse }>(
    `/projects/${projectIdentifier}/web-functions/sub-functions`,
    { params }
  );
  return response.data;
}

/**
 * 获取 Web 子功能详情
 */
export async function getWebSubFunction(
  projectIdentifier: string,
  subFunctionId: string
): Promise<WebSubFunction> {
  const response = await apiClient.get<{ data: WebSubFunction }>(
    `/projects/${projectIdentifier}/web-functions/sub-functions/${subFunctionId}`
  );
  return response.data;
}
// FIXME  My80OmFIVnBZMlhsdktEbHVwNDZSbkZxYVE9PTo0N2E2MDQ1ZA==

/**
 * 创建 Web 子功能
 */
export async function createWebSubFunction(
  projectIdentifier: string,
  data: CreateWebSubFunctionRequest
): Promise<WebSubFunction> {
  const response = await apiClient.post<{ data: WebSubFunction }>(
    `/projects/${projectIdentifier}/web-functions/sub-functions`,
    data
  );
  return response.data;
}

/**
 * 更新 Web 子功能
 */
export async function updateWebSubFunction(
  projectIdentifier: string,
  subFunctionId: string,
  data: UpdateWebSubFunctionRequest
): Promise<WebSubFunction> {
  const response = await apiClient.patch<{ data: WebSubFunction }>(
    `/projects/${projectIdentifier}/web-functions/sub-functions/${subFunctionId}`,
    data
  );
  return response.data;
}

/**
 * 删除 Web 子功能
 */
export async function deleteWebSubFunction(
  projectIdentifier: string,
  subFunctionId: string
): Promise<{ success: boolean; message: string }> {
  const response = await apiClient.delete<{ data: { success: boolean; message: string } }>(
    `/projects/${projectIdentifier}/web-functions/sub-functions/${subFunctionId}`
  );
  return response.data;
}

/**
 * 获取文件夹下的 Web 功能
 */
export async function getFolderWebFunctions(
  projectIdentifier: string,
  folderId: string,
  params?: {
    offset?: number;
    limit?: number;
    search?: string;
  }
): Promise<WebFunctionListResponse> {
  const response = await apiClient.get<{ data: WebFunctionListResponse }>(
    `/projects/${projectIdentifier}/web-functions`,
    {
      params: {
        ...params,
        folder_id: folderId,
      },
    }
  );
  return response.data;
}

/**
 * 获取文件夹下的 Web 子功能
 */
export async function getFolderWebSubFunctions(
  projectIdentifier: string,
  folderId: string,
  params?: {
    offset?: number;
    limit?: number;
    search?: string;
  }
): Promise<WebSubFunctionListResponse> {
  const response = await apiClient.get<{ data: WebSubFunctionListResponse }>(
    `/projects/${projectIdentifier}/web-functions/sub-functions`,
    {
      params: {
        ...params,
        folder_id: folderId,
      },
    }
  );
  return response.data;
}
